### PR TITLE
For 1.18 analysis direct tps

### PR DESCRIPTION
**Pull request by DiyiWi**

### CODE REVIEW

File: internal/adapt/direct_adapter/handler.go [hunk 1]
```diff
@@ -35,27 +42,88 @@ func (d *DirectAdapter) handleGetIBTPMessage(stream network.Stream, msg *pb.Mess
...
+							d.maxIndexReceiptMap.Store(servicePair, ibtp.Index)
+						}
+
+						if ibtp.Index < load.(uint64)+1 {
+							continue
+						}
+
+						if ibtp.Index > load.(uint64)+1 {
+							pool.resp_ibtps.Store(ibtp.Index, ibtp)
+							for i := load.(uint64) + 1; i <= ibtp.Index; i++ {
```
Comment (jzhe886) [inline internal/adapt/direct_adapter/handler.go:111]: where delete？

File: internal/adapt/direct_adapter/handler.go [hunk 2]
```diff
@@ -35,27 +42,88 @@ func (d *DirectAdapter) handleGetIBTPMessage(stream network.Stream, msg *pb.Mess
...
+							d.maxIndexMap.Store(servicePair, ibtp.Index)
+						}
+
+						if ibtp.Index < load.(uint64)+1 {
+							continue
+						}
+
+						if ibtp.Index > load.(uint64)+1 {
+							pool.req_ibtps.Store(ibtp.Index, ibtp)
+							for i := load.(uint64) + 1; i <= ibtp.Index; i++ {
```
Comment (jzhe886) [inline internal/adapt/direct_adapter/handler.go:88]: where delete ibtp？

File: internal/adapt/direct_adapter/gopool.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+func (p *pool) Add(delta int) {
+	for i := 0; i < delta; i++ {
+		p.queue <- 1
+	}
+	for i := 0; i > delta; i-- {
+		<-p.queue
+	}
+	p.wg.Add(delta)
+}
```
Comment (jzhe886) [inline internal/adapt/direct_adapter/gopool.go:30]: 1. blocked when delta > queue. size;
2. blocked when delta is negative number;

File: internal/adapt/direct_adapter/gopool.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+		size = 1
+	}
+	return &pool{
+		queue: make(chan int, size),
+		wg:    &sync.WaitGroup{},
+	}
+}
+
+func (p *pool) Add() {
+	p.queue <- 1
```
Comment (jzhe886) [inline internal/adapt/direct_adapter/gopool.go:23]: push struct{}{}

**Review by Karenlrx — APPROVED** [review]

**Pull request merged**